server/lib/nl/detection/date.py

```diff
@@ -115,30 +116,60 @@ def get_date_string(date: Date) -> str:
   if not date or not date.year:
     return ''
   year_string = str(date.year)
-  month_string = _get_month_string(date)
+  month_string = _get_month_string(date.month)
   return year_string + month_string
 
 
+# Gets the year and month to use as the base date when calculating a date range
+def _get_base_year_month(date: Date) -> (int, int):
+  base_year = date.year
+  base_month = date.month
+  # if date range excludes the specified date, need to do some calculations to
+  # get the base date.
+  if date.prep in _EXCLUSIVE_DATE_PREPS:
+    # if specified date is an end date, base date should be earlier than
+    # specified date
+    if date.prep in _END_DATE_PREPS:
+      # if date is monthly, use date that is one month before the specified date
+      if date.month >= _MIN_MONTH:
+        base_year = base_year - 1 if date.month == _MIN_MONTH else base_year
+        base_month = _MAX_MONTH if date.month == _MIN_MONTH else base_month - 1
```

Comment:
Would something like dateutil (which we already use) make it easier without scope for off-by-ones?  https://pypi.org/project/python-dateutil/

Reply:
good idea, updated!